File: rust/tests/differential.rs
//! Nightly differential conformance against a randomized C-reference corpus.

use std::{env, fs};

fn take<'a>(corpus: &'a [u8], offset: &mut usize, len: usize) -> &'a [u8] {
    let end = offset
        .checked_add(len)
        .filter(|&end| end <= corpus.len())
        .expect("truncated differential corpus");
    let value = &corpus[*offset..end];
    *offset = end;
    value
}

fn read_u32(corpus: &[u8], offset: &mut usize) -> u32 {
    u32::from_le_bytes(take(corpus, offset, 4).try_into().unwrap())
}

fn read_u64(corpus: &[u8], offset: &mut usize) -> u64 {
    u64::from_le_bytes(take(corpus, offset, 8).try_into().unwrap())
}

#[test]
fn randomized_c_reference_corpus() {
    let Ok(path) = env::var("HAYAHASH_CORPUS") else {
        eprintln!("HAYAHASH_CORPUS is unset; skipping nightly differential corpus");
        return;
    };
    let corpus = fs::read(&path).unwrap_or_else(|err| panic!("cannot read {path}: {err}"));
    let mut offset = 0;
    assert_eq!(take(&corpus, &mut offset, 8), b"HAYAFZ01");
    let case_count = read_u32(&corpus, &mut offset);
    let prng_seed = read_u64(&corpus, &mut offset);

    for case_index in 0..case_count {
        let len = read_u32(&corpus, &mut offset) as usize;
        let hash_seed = read_u64(&corpus, &mut offset);
        let expected = read_u64(&corpus, &mut offset);
        let input = take(&corpus, &mut offset, len);
        let actual = hayahash::hayahash64(input, hash_seed);
        assert_eq!(
            actual, expected,
            "case={case_index} len={len} hash_seed={hash_seed:#018x} \
             corpus_prng_seed={prng_seed:#018x}"
        );
    }

    assert_eq!(
        offset,
        corpus.len(),
        "trailing bytes in differential corpus"
    );
    eprintln!(
        "Rust matched {case_count} C-reference cases \
         (corpus PRNG seed={prng_seed:#018x})"
    );
}

File: zig/tests/differential.zig
//! Nightly differential conformance against a randomized C-reference corpus.

const std = @import("std");
const hayahash = @import("hayahash");

const CorpusReader = struct {
    bytes: []const u8,
    offset: usize = 0,

    fn take(self: *CorpusReader, len: usize) ![]const u8 {
        const end = std.math.add(usize, self.offset, len) catch
            return error.InvalidDifferentialCorpus;
        if (end > self.bytes.len) return error.InvalidDifferentialCorpus;
        defer self.offset = end;
        return self.bytes[self.offset..end];
    }

    fn readU32(self: *CorpusReader) !u32 {
        const bytes = try self.take(4);
        return std.mem.readInt(u32, bytes[0..4], .little);
    }

    fn readU64(self: *CorpusReader) !u64 {
        const bytes = try self.take(8);
        return std.mem.readInt(u64, bytes[0..8], .little);
    }
};

test "randomized C-reference corpus" {
    const path = std.testing.environ.getAlloc(
        std.testing.allocator,
        "HAYAHASH_CORPUS",
    ) catch |err| switch (err) {
        error.EnvironmentVariableMissing => return,
        else => return err,
    };
    defer std.testing.allocator.free(path);

    const corpus = try std.Io.Dir.cwd().readFileAlloc(
        std.testing.io,
        path,
        std.testing.allocator,
        .limited(512 * 1024 * 1024),
    );
    defer std.testing.allocator.free(corpus);

    var reader: CorpusReader = .{ .bytes = corpus };
    try std.testing.expectEqualStrings("HAYAFZ02", try reader.take(8));
    const case_count = try reader.readU32();
    const prng_seed = try reader.readU64();
    for (0..case_count) |case_index| {
        const len: usize = try reader.readU32();
        const hash_seed = try reader.readU64();
        const expected_lo = try reader.readU64();
        const expected_hi = try reader.readU64();
        const input = try reader.take(len);
        const actual = hayahash.hayahash128(input, hash_seed);
        if (actual.lo != expected_lo or actual.hi != expected_hi) {
            std.debug.print(
                "case={d} len={d} hash_seed=0x{x:0>16} " ++
                    "corpus_prng_seed=0x{x:0>16}\n",
                .{ case_index, len, hash_seed, prng_seed },
            );
            try std.testing.expectEqual(expected_lo, actual.lo);
            try std.testing.expectEqual(expected_hi, actual.hi);
        }
        try std.testing.expectEqual(hayahash.hayahash64(input, hash_seed), actual.lo);
    }

    try std.testing.expectEqual(corpus.len, reader.offset);
}
